apps/ios/Shared/Views/Database/DatabaseEncryptionView.swift

```diff
@@ -0,0 +1,286 @@
+        }
+    }
+}
+
+struct DatabaseEncryptionView: View {
+    @EnvironmentObject private var m: ChatModel
+    @State private var alert: DatabaseEncryptionAlert? = nil
+    @State private var progressIndicator = false
+    @State private var useKeychain = storeDBPassphraseGroupDefault.get()
+    @State private var initialRandomDBPassphrase = initialRandomDBPassphraseGroupDefault.get()
+    @State private var storedKey = getDatabaseKey() != nil
+    @State private var currentKey = ""
+    @State private var newKey = ""
+    @State private var confirmNewKey = ""
+    @State private var currentKeyShown = false
+
+    var body: some View {
+        ZStack {
+            databaseEncryptionView()
+            if progressIndicator {
+                ProgressView().scaleEffect(2)
+            }
+        }
+    }
+
+    private func databaseEncryptionView() -> some View {
+        List {
+            Section {
+                settingsRow("key") {
+                    Toggle("Save passphrase in Keychain", isOn: $useKeychain)
+                    .onChange(of: useKeychain) { _ in
+                        if useKeychain {
+                            storeDBPassphraseGroupDefault.set(true)
+                        } else if storedKey {
+                            alert = .keychainRemoveKey
+                        } else {
+                            storeDBPassphraseGroupDefault.set(false)
+                        }
+                    }
+                    .disabled(initialRandomDBPassphrase)
+                }
+
+                if !initialRandomDBPassphrase && m.chatDbEncrypted == true {
+                    DatabaseKeyField(key: $currentKey, placeholder: "Current passphrase…", valid: validKey(currentKey))
+                }
+
+                DatabaseKeyField(key: $newKey, placeholder: "New passphrase…", valid: validKey(newKey))
+                DatabaseKeyField(key: $confirmNewKey, placeholder: "Confirm new passphrase…", valid: confirmNewKey == "" || newKey == confirmNewKey)
+
+                settingsRow("lock.rotation") {
+                    Button("Update database passphrase") {
+                        alert = currentKey == ""
+                        ? (useKeychain ? .encryptDatabaseSaved : .encryptDatabase)
+                        : (useKeychain ? .changeDatabaseKeySaved : .changeDatabaseKey)
+                    }
+                }
+                .disabled(
+                    currentKey == newKey ||
+                    newKey != confirmNewKey ||
+                    newKey == "" ||
+                    !validKey(currentKey) ||
+                    !validKey(newKey)
+                )
+            } header: {
+                Text("")
+            } footer: {
+                VStack(alignment: .leading, spacing: 16) {
+                    if m.chatDbEncrypted == false {
+                        Text("Your chat database is not encrypted - set passphrase to encrypt it.")
+                    } else if useKeychain {
```

Comment:
use separate variable for determining text to show, so that when disabling keychain text doesn't change behind alert before confirmation? (the variable to be changed on confirm)